pydantic/_internal/_generate_schema.py

```diff
@@ -757,10 +757,14 @@ def set_discriminator(schema: CoreSchema) -> CoreSchema:
             'examples': field_info.examples,
         }
         json_schema_updates = {k: v for k, v in json_schema_updates.items() if v is not None}
-        json_schema_updates.update(field_info.json_schema_extra or {})
+        if isinstance(field_info.json_schema_extra, dict):
+            json_schema_updates.update(field_info.json_schema_extra)
 
         def json_schema_update_func(schema: CoreSchemaOrField, handler: GetJsonSchemaHandler) -> JsonSchemaValue:
-            return {**handler(schema), **json_schema_updates}
+            json_schema = {**handler(schema), **json_schema_updates}
+            if callable(field_info.json_schema_extra):
+                field_info.json_schema_extra(json_schema)
```

Comment:
again, I'm a bit lost why we don't update do `json_schema.update(field_info.json_schema_extra)` here if it's a dict?

If you sure these are fine, great.

Reply:
I think we can clean this up, I'll push a commit doing that shortly